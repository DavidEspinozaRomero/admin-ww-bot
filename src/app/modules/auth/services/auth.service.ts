import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { catchError, map, of, tap } from 'rxjs';

import { environment } from 'src/environments/environment';
import { StorageService } from '../../shared/services/storage.service';
import { LocalStorageKey } from 'src/app/modules/shared/interfaces/storage.interface';
import {
  authloginResponse,
  registerUser,
  loginUser,
  User,
} from '../interfaces/auth.interface';

@Injectable({ providedIn: 'root' })
export class AuthService {
  private readonly BaseUrl = environment.UrlApi;
  private _user?: User;

  public get user() {
    return { ...this._user };
  }

  constructor(
    private readonly httpClient: HttpClient,
    private readonly storage: StorageService
  ) {}

  registerUser(json: registerUser) {
    const URL = `${this.BaseUrl}auth/register`;
    return this.httpClient.post<any>(URL, json);
  }
  loginUser(json: loginUser) {
    const URL = `${this.BaseUrl}auth/login`;
    return this.httpClient.post<authloginResponse>(URL, json).pipe(
      tap((res) => {
        this.setUserAndToken(res);
      })
    );
  }

  validarToken() {
    const URL = `${this.BaseUrl}auth/check-status`;
    const headers: HttpHeaders = new HttpHeaders().set(
      'Authorization',
      `Bearer ${this.storage.getLocalStorage(LocalStorageKey.token) || ''}`
    );
    return this.httpClient.get<authloginResponse>(URL, { headers }).pipe(
      map((res) => {
        this.setUserAndToken(res);
        return true;
      }),
      catchError((err) => of(false))
    );
  }

  //#region method
  setUserAndToken(res: authloginResponse) {
    const { message, token, ...userData } = res;
    this.storage.setLocalStorage(LocalStorageKey.token, token);
    this._user = userData;
  }
  //#endregion method

  //#region
  //#endregion
}
