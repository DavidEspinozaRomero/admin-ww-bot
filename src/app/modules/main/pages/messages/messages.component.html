<!-- <p>settings works!</p>
<span> user: </span>
<pre> {{ user | json }} </pre> -->

<h1>ws conect</h1>
<span> {{ status.conected ? "conect" : "aoeu" }} </span>

<section>
  <button
    #BtnNewMessage
    type="button"
    class="btn btn-primary mb-3"
    data-bs-toggle="collapse"
    data-bs-target="#collapseExample"
    aria-expanded="false"
    aria-controls="collapseExample"
  >
    + Nuevo Mensaje
  </button>
  <!-- (click)="status.showFormMessage = true" -->

  <!-- *ngIf="status.showFormMessage" -->
  <form
    class="mb-3 collapse"
    id="collapseExample"
    [formGroup]="newMessageGroup"
  >
    <div class="row">
      <div class="col">
        <div class="mb-3">
          <label for="message" class="form-label">Mensage</label>
          <textarea
            type="text"
            class="form-control form-control-sm"
            name="message"
            id=""
            aria-describedby="helpId"
            placeholder=" -- Agregue el mensaje --"
            formControlName="message"
            rows="4"
            [ngClass]="{ 'is-invalid': isValid(newMessageGroup, 'message') }"
          >
          </textarea>
          <!-- <small id="helpId" class="form-text text-muted">Help text</small> -->
        </div>
      </div>
      <div class="col-4">
        <div class="mb-1">
          <label for="category" class="form-label">Categoria</label>
          <select
            class="form-select form-select-sm"
            name="category"
            id=""
            formControlName="category"
            [ngClass]="{'is-invalid': isValid(newMessageGroup, 'category')}"
          >
            <option value="" selected>-- Select one --</option>
            <option [value]="item" *ngFor="let item of categories">
              {{ item | titlecase }}
            </option>
          </select>
        </div>

        <div class="mb-3">
          <label for="date" class="form-label">Fecha</label>
          <input
            type="date"
            class="form-control form-control-sm"
            name="date"
            id=""
            aria-describedby="helpId"
            placeholder=" -- Agregue el mensaje --"
            formControlName="date"
          />
          <small
            *ngIf="isValid(newMessageGroup, 'date')"
            id="helpId"
            class="form-text text-muted"
            >Help text</small
          >
        </div>
        <!-- <div class="mb-3">
              <label for="asdf" class="form-label">Hora</label>
              <input
                type="time"
                class="form-control form-control-sm"
                name="asdf"
                id=""
                aria-describedby="helpId"
                placeholder=" -- Agregue el mensaje --"
              />
            </div> -->
      </div>
    </div>
    <div class="text-end">
      <!-- (click)="status.showFormMessage = false" -->
      <button *ngIf="!status.spinner" type="button" class="btn btn-primary me-3" (click)="save()">
        Guardar
      </button>
      <button *ngIf="status.spinner" class="btn btn-primary" type="button" disabled>
        <span class="spinner-border spinner-border-sm" role="status" aria-hidden="true"></span>
        Guardando...
      </button>
      <!-- (click)="status.showFormMessage = false" -->
      <button type="button" class="btn btn-outline-danger" (click)="cancel()">
        Cancelar
      </button>
    </div>
  </form>
  <!-- <pre> {{ newMessageGroup.value | json }} </pre> -->

  <!-- TODO: agregar lista de mensajes -->
  <div class="table-responsive">
    <table
      class="table table-striped table-hover table-borderless align-middle"
    >
      <thead class="table-primary text-primary">
        <!-- <caption>
          Table Name
        </caption> -->
        <tr>
          <th class="" style="min-width: 10rem">Mensaje</th>
          <th class="text-center" style="width: 8rem">Categoria</th>
          <th class="" style="width: 8rem">Fecha</th>
          <th class="" style="width: 5rem"></th>
        </tr>
      </thead>
      <tbody class="table-group-divider">
        <tr *ngFor="let msg of listMessages; index as i_msg" class="">
          <td scope="row"> {{msg.message}} </td>
          <td class="text-center">
            <span class="badge bg-primary"> {{msg.category}} </span>
          </td>
          <td>
            {{ msg.date | date: "mediumDate" }}
          </td>
          <td>
            <i
              class="bi bi-pencil-square bg-white border rouded px-1 me-2"
              role="button"
              (click)="edit(msg, i_msg)"
            ></i>
            <i
              class="bi bi-trash text-danger bg-white border rouded px-1"
              role="button"
              (click)="erase(msg, i_msg)"
            ></i>
          </td>
        </tr>
        <!-- <tr class="table-primary">
          <td scope="row">Item</td>
          <td>Item</td>
          <td>Item</td>
        </tr> -->
      </tbody>
      <tfoot></tfoot>
    </table>
  </div>
</section>

<!-- <section class="row justify-content-center">
  <div class="col-4">
    <h2>Scan de QRCODE to use the app</h2>
    <ol>
      <li>Abre WhatsApp en tu teléfono</li>
      <li>
        Toca Menu <i class="bi bi-three-dots-vertical"></i> o Configuracion
        <i class="bi bi-gear"></i> y selecciona dispositivos vinculados
      </li>
      <li>
        Cuando se active la cámara, apunta tu teléfono hacia esta pantalla para
        escanear el código
      </li>
    </ol>
  </div>
  <div class="col-3" style="background-color: #eee">
    <div
      *ngIf="status.loadingQR"
      class="d-flex h-100 justify-content-center align-items-center"
    >
      <div class="spinner-border" role="status">
        <span class="visually-hidden">Loading...</span>
      </div>
    </div>
    <div
      *ngIf="!status.loadingQR && !status.responseQR"
      class="h-100 text-center p-3"
    >
      <p class="mt-4">
        <i class="bi bi-arrow-clockwise fs-1 text-success"></i>
      </p>
      <button class="btn btn-outline-primary" (click)="getQRCODE()">
        Refresh code
      </button>
    </div>
    <div #qrcode *ngIf="!status.loadingQR && status.responseQR" class=""></div>
  </div>
</section> -->
