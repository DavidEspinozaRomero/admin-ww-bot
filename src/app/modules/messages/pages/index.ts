export { MessagesComponent } from "./messages/messages.component";

