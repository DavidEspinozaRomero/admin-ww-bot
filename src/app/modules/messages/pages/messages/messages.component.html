<app-loading *ngIf="status.loading"></app-loading>
<app-retry
  *ngIf="!status.loading && !status.response"
  (onRetry)="ngOnInit()"
></app-retry>

<span>
  {{ status.conected ? "Conectado" : "Desconectado" }}
  <i
    class="rounded-circle d-inline-block"
    style="width: 0.5rem; height: 0.5rem"
    [ngClass]="{
      'bg-success': status.conected,
      'bg-danger': !status.conected
    }"
  >
  </i>
</span>

<section *ngIf="!status.loading && status.response">
  <div class="row">
    <div class="col-auto">
      <button
        #BtnNewMessage
        type="button"
        class="btn btn-primary mb-3"
        data-bs-toggle="collapse"
        data-bs-target="#collapseExample"
        aria-expanded="false"
        aria-controls="collapseExample"
      >
        + New Message
      </button>
    </div>
    <div class="col-auto">
      <button
        #modalqrcode
        type="button"
        class="btn btn-primary"
        data-bs-toggle="modal"
        data-bs-target="#modalqrcode"
      >
        Add WhatsAppWeb
      </button>
    </div>
  </div>

  <form class="mb-3 collapse" id="collapseExample" [formGroup]="messageForm">
    <div class="row">
      <div class="col">
        <div class="mb-3">
          <label for="keyword" class="form-label">Keyword</label>
          <div class="input-group">
            <input
              type="text"
              class="form-control form-control-sm"
              id="keyword"
              aria-describedby="helpId"
              placeholder=" -- Agregue el mensaje --"
              [formControl]="keywordControl"
              [ngClass]="{ 'is-invalid': utils.isValid(keywordControl) }"
            />
            <i
              class="input-group-text bg-primary text-white"
              (click)="addKeyword()"
              >+</i
            >
          </div>
          <small
            *ngIf="utils.isValid(keywordControl)"
            id="helpId"
            class="form-text"
            err-msg
            >min 2 letters</small
          >

          <div>
            <span
              *ngFor="let keyword of keywordsArr; let i_keyword = index"
              class="badge bg-primary me-1"
            >
              {{ keyword }}
              <i
                class="bi bi-x-circle text-danger"
                (click)="removeKeyword(i_keyword)"
              ></i>
            </span>
          </div>
          <small
            *ngIf="utils.isValidControl(messageForm, 'keywords')"
            id="helpId"
            class="form-text"
            err-msg
            >Need 1 keyword min</small
          >
        </div>
        <div class="mb-3">
          <label for="answer" class="form-label">Response</label>
          <textarea
            type="text"
            class="form-control form-control-sm"
            id="answer"
            aria-describedby="helpId"
            placeholder=" -- Agregue el mensaje --"
            formControlName="answer"
            rows="4"
            [ngClass]="{
              'is-invalid': utils.isValidControl(messageForm, 'answer')
            }"
          >
          </textarea>
          <small
            *ngIf="utils.isValidControl(messageForm, 'answer')"
            id="helpId"
            class="form-text text-muted"
            >Add response</small
          >
        </div>
      </div>
      <div class="col-4">
        <div class="mb-3">
          <label for="category" class="form-label">Category</label>
          <select
            class="form-select form-select-sm"
            id="category"
            formControlName="category"
            [ngClass]="{
              'is-invalid': utils.isValidControl(messageForm, 'category')
            }"
          >
            <option value="" selected>-- Select one --</option>
            <option [value]="category.id" *ngFor="let category of categories">
              {{ category.description | titlecase }}
            </option>
          </select>
        </div>

        <div class="mb-1">
          <label for="startTime" class="form-label">Start time</label>
          <input
            type="time"
            class="form-control form-control-sm"
            id="startTime"
            aria-describedby="helpId"
            formControlName="startTime"
          />
          <!-- <small
            *ngIf="utils.isValidControl(messageForm, 'type')"
            id="helpId"
            class="form-text text-muted"
          >
            Help text
          </small> -->
        </div>
        <div class="mb-1">
          <label for="endTime" class="form-label">End time</label>
          <input
            type="time"
            class="form-control form-control-sm"
            id="endTime"
            aria-describedby="helpId"
            formControlName="endTime"
          />
          <!-- <small
            *ngIf="utils.isValidControl(messageForm, 'type')"
            id="helpId"
            class="form-text text-muted"
          >
            Help text
          </small> -->
        </div>
      </div>
    </div>

    <div class="text-end">
      <button
        *ngIf="!status.spinner"
        type="button"
        class="btn btn-primary me-3"
        (click)="save()"
      >
        Save
      </button>
      <button
        *ngIf="status.spinner"
        class="btn btn-primary"
        type="button"
        disabled
      >
        <span
          class="spinner-border spinner-border-sm"
          role="status"
          aria-hidden="true"
        ></span>
        Saving...
      </button>
      <button type="button" class="btn btn-outline-danger" (click)="cancel()">
        Cancel
      </button>
    </div>
  </form>

  <!-- <pre> {{messageForm.value | json}} </pre> -->
  <div>
    <ul class="nav nav-tabs mb-3">
      <li
        *ngFor="let item of menu"
        class="nav-item"
        [attr.aria-label]="item.label"
        (click)="selecNavItem(item)"
      >
        <a
          role="button"
          class="nav-link"
          [ngClass]="{ 'active bg-primary text-white ': item.selected }"
        >
          {{ item.label | titlecase }}
        </a>
      </li>
    </ul>
  </div>

  <section *ngIf="menu[0].selected" class="table-responsive">
    <table
      aria-label="questions-answers"
      class="table table-striped table-hover table-borderless align-middle"
    >
      <thead class="table-primary text-primary">
        <tr>
          <th class="">Keywords</th>
          <th class="">Response</th>
          <th class="text-center">Category</th>
          <th class="">Time</th>
          <th class="">Actions</th>
        </tr>
      </thead>
      <tbody class="table-group-divider">
        <tr *ngFor="let msg of listMessages; index as i_msg" class="">
          <td>
            <span *ngFor="let keyword of msg.keywords; index as i_tkeyword">
              {{ keyword }}
              {{ i_tkeyword + 1 < msg.keywords.length ? "/" : null }}
            </span>
          </td>
          <td>{{ msg.answer }}</td>
          <td class="text-center">
            <span class="badge bg-primary"> {{ msg.category }} </span>
          </td>
          <td>
            {{ msg.startTime }} -
            {{ msg.endTime }}
          </td>
          <td>
            <i
              class="bi bi-pencil-square bg-white border rouded px-1 me-2"
              role="button"
              aria-label="btn-edit"
              (click)="edit(msg, i_msg)"
            ></i>
            <i
              class="bi bi-trash text-danger bg-white border rouded px-1"
              role="button"
              aria-label="btn-delete"
              (click)="erase(msg, i_msg)"
            ></i>
          </td>
        </tr>
      </tbody>
      <tfoot></tfoot>
    </table>
  </section>

  <div *ngIf="menu[1].selected">
    <app-in-progress></app-in-progress>
  </div>
</section>

<app-modalqrcode [socket]="this.socket"></app-modalqrcode>

<!-- codigo para otro tipo de form message -->

<!-- <div class="mb-3">
  <div class="row">
    <div class="col">
      <label for="date" class="form-label">Fecha</label>
    </div>
    <div class="col">
      <div class="form-check form-switch">
        <input
          class="form-check-input"
          type="checkbox"
          role="switch"
          id="flexSwitchCheckDefault"
          formControlName="type"
        />
        <label class="form-check-label" for="flexSwitchCheckDefault">
          {{ messageForm.value.type ? "Day" : "Hour" }}
        </label>
      </div>
    </div>
  </div>

  <input
    type="time"
    class="form-control form-control-sm"
    name="startTime"
    id=""
    aria-describedby="helpId"
    formControlName="startTime"
  />
  <small
    *ngIf="utils.isValidControl(messageForm, 'type')"
    id="helpId"
    class="form-text text-muted"
  >
    Help text
  </small>
</div> -->
