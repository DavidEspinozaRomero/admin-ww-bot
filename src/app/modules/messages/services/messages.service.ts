import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';

import { StorageService } from '../../../services';
import { environment } from '../../../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class MessagesService {
  private readonly BaseUrl = environment.UrlApi;

  constructor(
    private readonly httpClient: HttpClient,
    private readonly storage: StorageService
  ) {}

  getqrimg() {
    const URL = `${this.BaseUrl}bot-webwhatsap/qrcode`;
    const headers: HttpHeaders = new HttpHeaders();
    headers.set('Accept', `image/svg+xml`);
    return this.httpClient.get(URL, { headers,  responseType: 'text' },);
  }

  createMessage(body: {}) {
    const URL = `${this.BaseUrl}messages`;
    return this.httpClient.post(URL, body);
  }

  getAllMessages() {
    const URL = `${this.BaseUrl}messages`;
    return this.httpClient.get(URL);
  }

  updateMessage(id: string, body: {}) {
    const URL = `${this.BaseUrl}messages/${id}`;
    return this.httpClient.patch(URL, body);
  }

  deleteMessage(id: string) {
    const URL = `${this.BaseUrl}messages/${id}`;
    return this.httpClient.delete(URL);
  }

  getCategories() {
    const URL = `${this.BaseUrl}messages/get-categories`;
    return this.httpClient.get(URL);
  }
}

