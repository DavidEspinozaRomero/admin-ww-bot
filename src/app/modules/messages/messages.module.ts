import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';

import { MessagesRoutingModule } from './messages-routing.module';
import { SharedModule } from '../shared/shared.module';

import { ModalqrcodeComponent } from './components/modalqrcode/modalqrcode.component';
import { MessagesComponent } from './pages/messages/messages.component';
import {
  InProgressModule,
  LoadingModule,
  RetryModule,
} from '../shared/components';

@NgModule({
  declarations: [ModalqrcodeComponent, MessagesComponent],
  imports: [
    CommonModule,
    ReactiveFormsModule,
    MessagesRoutingModule,
    SharedModule,
    LoadingModule,
    InProgressModule,
    RetryModule,
  ],
})
export class MessagesModule {}
