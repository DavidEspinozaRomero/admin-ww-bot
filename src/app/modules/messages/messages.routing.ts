import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';

import { MessagesComponent } from './pages/messages/messages.component';

const routes: Routes = [
  { path: '', component: MessagesComponent },
  { path: '**', redirectTo: '' },
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class MessagesRoutingModule { }
