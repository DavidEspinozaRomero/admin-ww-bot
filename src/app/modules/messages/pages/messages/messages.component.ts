import { Component, OnInit, ViewChild, ElementRef } from '@angular/core';
import { Manager, Socket } from 'socket.io-client';
import {
  FormBuilder,
  FormControl,
  FormGroup,
  Validators,
} from '@angular/forms';

import { forkJoin } from 'rxjs';

import { CustomToastService } from '../../../../services';
import { MessagesService } from '../../services/messages.service';
import {
  LocalStorageKey,
  StorageService,
} from '../../../../services/storage.service';
import { Category, Message } from '../../interfaces/message.inteface';
import { environment } from '../../../../../environments/environment';


@Component({
  selector: 'app-messages',
  templateUrl: './messages.component.html',
  styleUrls: ['./messages.component.scss']
})
export class MessagesComponent implements OnInit {
  //#region
  //#endregion
  //#region variables
  @ViewChild('qrcode') qrcodeDiv!: ElementRef<HTMLDivElement>;
  @ViewChild('BtnNewMessage') BtnNewMessage?: ElementRef<HTMLButtonElement>;
  @ViewChild('formColapse') formColapse?: ElementRef<HTMLFormElement>;
  @ViewChild('modalqrcode') modalqrcode?: ElementRef<HTMLButtonElement>;

  status = {
    loading: false,
    response: false,
    spinner: false,
    conected: false,
    showFormMessage: false,
  };

  menu = [
    { label: 'respuestas', selected: true },
    { label: 'mensajes', selected: false },
  ];

  categories: Category[] = [];
  listMessages: any[] = [];

  msgIndex!: number;

  showCode: boolean = false;
  private URLWS: string = `${environment.UrlApi}${environment.WebSocketUrl}`;
  manager: any;
  socket!: Socket;

  newMessageGroup: FormGroup = this.fb.group({
    id: [null],
    query: [null, [Validators.required, Validators.minLength(2)]],
    answer: [null, [Validators.required, Validators.minLength(2)]],
    category: ['', Validators.required],
    startTime: [null, [Validators.required]],
    endTime: [null, [Validators.required]],
  });

  miForm: FormGroup = this.fb.group({
    messages: this.fb.array([]),
  });

  //#endregion variables

  constructor(
    private readonly fb: FormBuilder,
    private readonly toast: CustomToastService,
    private readonly storage: StorageService,
    private readonly messagesService: MessagesService
  ) {}

  ngOnInit(): void {
    this.wsConect();
    this.initApis();
  }

  //#region WebSocket
  wsConect() {
    this.manager = new Manager(this.URLWS, {
      extraHeaders: {
        authorization:
          this.storage.getLocalStorage(LocalStorageKey.token) || '',
      },
    });
    this.socket?.removeAllListeners();

    this.socket = this.manager.socket('/');

    this.socket.on('connect', () => {
      console.log('conect');
    });
    this.socket.on('disconnect', () => {
      console.log('disconect');
      this.status.conected = false;
    });

    this.socket.on(
      'message-from-server',
      (payload: { action: string; description: string }) => {
        const { action, description } = payload;
        if (action == 'ready') {
          this.status.conected = true;
        }
      }
    );
  }

  //#endregion WebSocket

  //#region Apis
  // Get
  // TODO: agregar evento para llamar a la api y traer el qrcode
  getAllMessages() {
    this.status.loading = true;
    this.messagesService
      .getAllMessages()
      .subscribe({
        next: (res) => {
          this.listMessages = res.data;
          this.status.response = true;
        },
        error: (err) => {
          this.toast.error(err.message);
          this.status.response = false;
        },
      })
      .add(() => (this.status.loading = false));
  }
  initApis() {
    forkJoin({
      allmessages: this.messagesService.getAllMessages(),
      categories: this.messagesService.getCategories(),
    })
      .subscribe({
        next: ({ allmessages, categories }: any) => {
          this.listMessages = allmessages.data as Message[];
          this.categories = categories;
          this.status.response = true;
        },
        error: (err) => {
          this.toast.error(err.message);
          this.status.response = false;
        },
      })
      .add(() => (this.status.loading = false));
  }

  // Post
  createMessage(json: {}) {
    this.messagesService.createMessage(json).subscribe({
      next: (message: any) => {
        message.category = this.replaceCategoryIdToDescription(
          message.category.id
        );
        this.listMessages.push(message);
        this.BtnNewMessage?.nativeElement.click();
        this.resetForm();
      },
      error: (err) => {
        this.toast.error(err.message);
      },
    });
  }

  // Update
  updateMessage(id: string, json: {}, index: number) {
    this.messagesService.updateMessage(id, json).subscribe({
      next: (res: any) => {
        const editedMessage = res.data;
        editedMessage.category = this.replaceCategoryIdToDescription(
          editedMessage.category.id
        );
        this.listMessages.splice(index, 1, editedMessage);
        this.BtnNewMessage?.nativeElement.click();
        this.resetForm();
      },
      error: (err) => {
        this.toast.error(err.message);
      },
    });
  }

  // Delete
  deleteMessage(id: string, cb: () => void) {
    this.messagesService.deleteMessage(id).subscribe({
      next: (res) => {
        cb();
      },
      error: (err) => {
        this.toast.error(err.message);
      },
    });
  }
  //#endregion Apis

  //#region methods
  selecNavItem(item: any) {
    this.menu.forEach((element) => {
      element.selected = false;
    });
    item.selected = true;
  }
  cancel() {
    this.BtnNewMessage?.nativeElement.click();
    this.resetForm();
  }

  edit(msg: any, index: number) {
    this.msgIndex = index;
    const newMessage = {
      ...msg,
      category: this.replaceCategoryDescriptionToId(msg.category),
    };
    this.newMessageGroup.reset(newMessage);
    const open = document
      .getElementById('collapseExample')
      ?.classList.contains('show');
    if (!open) {
      this.BtnNewMessage?.nativeElement.click();
    }
  }
  erase(msg: any, index: number) {
    this.deleteMessage(msg.id, () => {
      this.listMessages.splice(index, 1);
    });
  }
  resetForm() {
    this.newMessageGroup.reset({ category: '' });
  }

  save() {
    this.newMessageGroup.markAllAsTouched();
    if (this.newMessageGroup.invalid) {
      this.toast.error('Falta llenar el formulario');
      return;
    }
    this.buildJson();
  }

  buildJson() {
    const {
      id = null,
      category = 1,
      ...data
    } = { ...this.newMessageGroup.value };
    if (id) {
      this.updateMessage(id, { category: +category, ...data }, this.msgIndex);
      return;
    }
    this.createMessage({ category: +category, ...data });
  }

  isValid(form: FormGroup, control: string) {
    return form.get(control)?.errors && form.get(control)?.touched;
  }

  replaceCategoryIdToDescription(id: number): string {
    return (
      this.categories.find((category) => category.id == id)?.description || ''
    );
  }

  replaceCategoryDescriptionToId(categoryDescription: string): number {
    return (
      this.categories.find(
        (category) => category.description == categoryDescription
      )?.id || 0
    );
  }

  //#endregion methods
}

//#region interfaces

//#endregion interfaces
