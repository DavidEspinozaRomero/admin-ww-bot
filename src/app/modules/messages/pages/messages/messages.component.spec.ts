import {
  HttpClientTestingModule,
  HttpTestingController,
} from '@angular/common/http/testing';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { FormBuilder, ReactiveFormsModule } from '@angular/forms';
import { render, RenderResult, screen } from '@testing-library/angular';
import { ToastrModule } from 'ngx-toastr';
import { of } from 'rxjs';

import { CustomToastService, StorageService } from '../../../../services';
import { MessagesService } from '../../services/messages.service';
import { MessagesComponent } from './messages.component';

//#region Mocks

const response = {
  qrimg: '<svg><svg/>',
  allmessages: {
    message: 'all ok',
    error: 'all bad',
    data: [
      {
        id: 1,
        query: 'hola',
        answer: 'hola david',
        category: '1',
        startTime: '8:00',
        endTime: '10:00',
      },
      {
        id: 2,
        query: 'hola2',
        answer: 'hola david2',
        category: '2',
        startTime: '9:00',
        endTime: '11:00',
      },
    ],
  },
  createmsg: {
    id: 1,
    query: 'hola',
    answer: 'hola david',
    category: 1,
    startTime: '8:00',
    endTime: '10:00',
  },
  updatemsg: {
    id: 1,
    query: 'saludo',
    answer: 'hola david e',
    category: 1,
    startTime: '8:00',
    endTime: '10:00',
  },
  deletemsg: {
    message: 'mensanje eliminado',
  },
  categories: [
    {
      id: 1,
      description: 'default',
    },
    {
      id: 2,
      description: 'family',
    },
    {
      id: 3,
      description: 'job',
    },
  ],
};

// const messagesServiceMock = createMock(MessagesService);
// messagesServiceMock.getAllMessages = jest.fn(() => of());
// messagesServiceMock.getCategories = jest.fn(() => of());
// messagesServiceMock.updateMessage = jest.fn(()=> of())
// messagesServiceMock.createMessage = jest.fn(()=> of())
// messagesServiceMock.deleteMessage = jest.fn(()=> of())
//#endregion Mocks

describe('MessagesComponent', () => {
  let service: MessagesService;
  let httpMock: HttpTestingController;
  let rendered: RenderResult<MessagesComponent>;
  let fixture: ComponentFixture<MessagesComponent>;
  let component: MessagesComponent;
  let compiled: HTMLElement;

  beforeEach(async () => {
    rendered = await render(MessagesComponent, {
      imports: [
        HttpClientTestingModule,
        ReactiveFormsModule,
        ToastrModule.forRoot({
          timeOut: 3000,
          positionClass: 'toast-bottom-right',
          preventDuplicates: true,
        }),
      ],
      providers: [
        FormBuilder,
        CustomToastService,
        StorageService,
        MessagesService,
      ],
    });
  });

  beforeEach(() => {
    service = TestBed.inject(MessagesService);
    httpMock = TestBed.inject(HttpTestingController);
    fixture = rendered.fixture;
    component = rendered.fixture.componentInstance;
    compiled = rendered.fixture.nativeElement;
  });

  it('should exist', () => {
    expect(component).toBeTruthy();
  });
  it('should exist', () => {
    expect(compiled).toMatchSnapshot();
  });
  describe('Layout', () => {
    const initMock = () => {
      httpMock
        .expectOne('http://localhost:3000/messages')
        .flush(response.allmessages);
      httpMock
        .expectOne('http://localhost:3000/messages/get-categories')
        .flush(response.categories);
      fixture.detectChanges();
    };
    it('has a init correctly', () => {
      const testReq = httpMock.expectOne('http://localhost:3000/messages');
      const testReq2 = httpMock.expectOne(
        'http://localhost:3000/messages/get-categories'
      );
      expect(testReq.request.method).toBe('GET');
      expect(testReq2.request.method).toBe('GET');
    });
    it('has 2 btn whit text "+ Nuevo Mensaje", "Vincular WhatsApp"', async () => {
      initMock();

      const btn1 = await screen.getByRole('button', {
        name: '+ Nuevo Mensaje',
      });
      const btn2 = await screen.getByRole('button', {
        name: 'Vincular WhatsApp',
      });

      expect(btn1).toBeInTheDocument();
      expect(btn2).toBeInTheDocument();
    });
    it('has form whit inputs, textarea, date selector, buttons save/cancel', () => {
      initMock();

      const inputQuery = screen.getByLabelText('Pregunta');
      const inputAnswer = screen.getByLabelText('Respuesta');
      const inputCategory = screen.getByLabelText('Categoria');
      const inputHoraInicio = screen.getByLabelText('Hora inicio');
      const inputHoraFin = screen.getByLabelText('Hora fin');
      const guardar = screen.getByRole('button', { name: 'Guardar' });
      const cancelar = screen.getByRole('button', { name: 'Cancelar' });

      expect(inputQuery).toBeInTheDocument();
      expect(inputAnswer).toBeInTheDocument();
      expect(inputCategory).toBeInTheDocument();
      expect(inputHoraInicio).toBeInTheDocument();
      expect(inputHoraFin).toBeInTheDocument();
      expect(guardar).toBeInTheDocument();
      expect(cancelar).toBeInTheDocument();
    });
    it('has nav tabs', () => {
      initMock();

      const nav = screen.getByRole('list', { name: '' });
      const listitem1 = screen.getByRole('button', { name: 'Respuestas' });
      const listitem2 = screen.getByRole('button', { name: 'Mensajes' });

      expect(nav).toBeInTheDocument();
      expect(listitem1).toBeInTheDocument();
      expect(listitem2).toBeInTheDocument();
    });
    it('has table whit headers: Pregunta, Respuesta, Categoria, Rango de horas, Acciones', () => {
      initMock();

      const table = screen.getByRole('table', { name: 'questions-answers' });
      const colh1 = screen.getByRole('columnheader', { name: 'Pregunta' });
      const colh2 = screen.getByRole('columnheader', { name: 'Respuesta' });
      const colh3 = screen.getByRole('columnheader', { name: 'Categoria' });
      const colh4 = screen.getByRole('columnheader', {
        name: 'Rango de horas',
      });
      const colh5 = screen.getByRole('columnheader', { name: 'Acciones' });

      expect(table).toBeInTheDocument();
      expect(colh1).toBeInTheDocument();
      expect(colh2).toBeInTheDocument();
      expect(colh3).toBeInTheDocument();
      expect(colh4).toBeInTheDocument();
      expect(colh5).toBeInTheDocument();
    });
  });
});
