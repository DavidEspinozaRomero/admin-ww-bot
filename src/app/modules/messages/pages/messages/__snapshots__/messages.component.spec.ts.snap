// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessagesComponent should match the snapshot 1`] = `
<div
  id="root1"
>
  
  <app-retry />
  
  <span>
     Desconectado 
    <i
      class="rounded-circle d-inline-block bg-danger"
      style="width: 0.5rem; height: 0.5rem;"
    />
  </span>
  
  <app-modalqrcode />
</div>
`;
